### PR TITLE
Delegate compression to servlet container

### PR DESCRIPTION
**Pull request by basil**
As noted in https://github.com/jenkinsci/stapler/issues/130 and https://github.com/jenkinsci/stapler/issues/265, doing compression at the web framework (Stapler) layer rather than the servlet container (Jetty) layer is some long-standing technical debt. The servlet layer can do this more effectively at a lower maintenance cost, and might even support additional compression algorithms like Brotli in the future pending https://github.com/jetty/jetty.project/issues/2553. Delegating compression to the servlet layer reduces the amount of code we need to maintain and opens the potential for us to adopt new compression schemes in the future as they are added upstream.

This PR goes with two others, all of which must be released together:

- https://github.com/jenkinsci/jenkins/pull/9379
- https://github.com/jenkinsci/stapler/pull/557
- https://github.com/jenkinsci/winstone/pull/390

These three PRs deprecate compression in Stapler and migrate Jenkins core away from any deprecated APIs. The deprecated APIs are left for plugin compatibility but delegate to the non-compressed versions instead. If these PRs are accepted, I will eventually go through critical plugins and migrate them off the deprecated APIs once their baseline contains this set of changes.

On the Winstone side, a new `--compression` option is added, which accepts `gzip` and `none` arguments, defaulting to `gzip` to preserve existing behavior. When using `gzip` compression, the Jetty `GzipHandler` is enabled, which transparently handles `gzip` compression when the client supports it. Of note is that it doesn't compress error pages (and this behavior is not configurable):

https://github.com/jetty/jetty.project/blob/7df1d7ec284d320cd7be8e1218ec6de95e9c032a/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHttpOutputInterceptor.java#L133-L139

On the Stapler side, all compression functionality is deleted except for `StaplerResponse#getCompressedOutputStream`, `StaplerResponse#getCompressedWriter`, and the `<st:compress>` Jelly tag (and its Groovy view equivalent). These are all retained for plugin compatibility but no longer consumed by Jenkins core as of this set of changes. The remaining implementations simply delegate to their non-compressed equivalents.

Previously `org.kohsuke.stapler.compression.CompressionFilter` did double duty as both a compression filter and an uncaught exception filter. These two functions have been separated, with the uncaught exception filter logic being retained in `UncaughtExceptionFilter`. As mentioned previously, Jetty's `GzipHandler` does not compress error pages, which is a change in behavior. We think that error pages should be rare enough that this change in behavior is acceptable. Anyone who is serious about compression should probably disable compression in Jetty and set up an `nginx` proxy with native (rather than JVM) Brotli compression or something more sophisticated.

On the core side, we simply pull in all of the above changes and migrate away from any deprecated APIs.

Another change in behavior is that the test harness will now be testing without `gzip` compression, since its copy of Jetty doesn't have the Winstone changes mentioned above. This shouldn't matter in practice, as I'll confirm by running PCT. Eventually I'll enable `GzipHandler` in the test harness, but I didn't want to do this yet because the test harness is still usable on older cores without this set of changes, and I wasn't sure if attempted double-compression would cause any problems (and didn't find a simple way to reliably detect from within Stapler whether `GzipHandler` was in operation at a lower layer). Eventually, once we're sure that consumers have this set of changes, I plan to re-enable `GzipHandler` in the test harness to bring tests closer to a production setup. We're planning on updating the test harness to require Jetty 12 in the next few months, so that would be a good opportunity to do so.

### Testing done

Everything new has unit tests and ATH/PCT are passing. In addition I ran `java -jar jenkins.war` both with and without `gzip` compression and verified in my browser that responses were being sent appropriately with either compression option. I confirmed that unfortunately Jetty's native `gzip` support will not compress an error page, even when compression is enabled.

### Proposed changelog entries

* Update Stapler from 1880.vb_6d94a_3b_05db_ to 1881.vd39f3ee5c629 and Winstone-Jetty from 6.19 to 6.20 to let Jetty handle HTTP response compression. A new command-line option `--compression` can be used to disable compression if desired.
* Developer: Deprecate `StaplerResponse#getCompressedOutputStream`, `StaplerResponse#getCompressedWriter`, and `st:compress`. Use `StaplerResponse#getOutputStream`, `StaplerResponse#getWriter`, and `j:jelly` instead.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [x] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [x] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [x] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [x] Proper changelog labels are set so that the changelog can be generated automatically.
- [x] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [x] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

**Review by daniel-beck — APPROVED** [review]
I updated the proposed changelog entries to cover both users and developer changes.

<details><summary>usage-in-plugins is also clean</summary>

```
org.kohsuke.stapler.StaplerResponse#getCompressedOutputStream
org.kohsuke.stapler.StaplerResponse#getCompressedWriter
org.kohsuke.stapler.compression.UncaughtExceptionHandler
org.kohsuke.stapler.compression.CompressionFilter
org.kohsuke.stapler.compression.CompressionServletResponse
org.kohsuke.stapler.compression.FilterServletOutputStream
org.kohsuke.stapler.jelly.DefaultScriptInvoker
```

finds `DefaultScriptInvoker` in the obsolete-looking https://plugins.jenkins.io/read-only-configurations/ with irrelevant matches, as well as just a handful of plugins using now deprecated methods.

</details>

**Review by MarkEWaite — APPROVED** [review]

**Comment by basil** [comment]
> `usage-in-plugins` is also clean

Ah yes, I had checked this a few weeks back but forgot to mention it in the PR description. Yeah, there are a very small number of plugins using the now-deprecated APIs which I can clean up after their baseline includes these changes.

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
